core/templates/pages/learner-dashboard-page/new-learner-dashboard-page/goals-tab.component.ts

```diff
@@ -0,0 +1,144 @@
+    private learnerDashboardIdsBackendApiService:
+      LearnerDashboardIdsBackendApiService) {
+  }
+  @Input() currentGoals: LearnerTopicSummary[];
+  @Input() editGoals: LearnerTopicSummary[];
+  @Input() completedGoals: LearnerTopicSummary[];
+  @Input() learntToPartiallyLearntTopics: string[];
+  learnerDashboardActivityIds: LearnerDashboardActivityIds;
+  MAX_CURRENT_GOALS_LENGTH: number;
+  currentGoalsStoryIsShown: boolean[];
+  topicBelongToCurrentGoals: boolean[] = [];
+  topicIdsInCompletedGoals: string[] = [];
+  topicIdsInEditGoals: string[] = [];
+  topicIdsInCurrentGoals: string[] = [];
+  topicMapping = {
+    'belongs to current goals': 0,
+    'belongs to completed goals': 1,
+    'belongs to neither': 2
+  };
+  activityType: string = constants.ACTIVITY_TYPE_LEARN_TOPIC;
+
+  ngOnInit(): void {
+    this.learnerDashboardIdsBackendApiService.
+      fetchLearnerDashboardIdsAsync().then(
+        (learnerDashboardActivityIds) => {
+          this.topicIdsInCompletedGoals = (
+            learnerDashboardActivityIds.learntTopicIds);
+          this.topicIdsInEditGoals = learnerDashboardActivityIds.allTopicIds;
+          this.topicIdsInCurrentGoals = (
+            learnerDashboardActivityIds.topicIdsToLearn);
+        }
+      );
+    this.MAX_CURRENT_GOALS_LENGTH = constants.MAX_CURRENT_GOALS_COUNT;
+    this.currentGoalsStoryIsShown = [];
+  }
+
+  getTopicPageUrl(
+      topicUrlFragment: string, classroomUrlFragment: string): string {
+    return this.urlInterpolationService.interpolateUrl(
+      ClassroomDomainConstants.TOPIC_VIEWER_URL_TEMPLATE, {
+        topic_url_fragment: topicUrlFragment,
+        classroom_url_fragment: classroomUrlFragment
+      });
+  }
+
+  topicBelongsTo(topicId: string): number {
+    if (this.topicIdsInCurrentGoals.includes(topicId)) {
+      return this.topicMapping['belongs to current goals'];
+    } else if (this.topicIdsInCompletedGoals.includes(topicId)) {
+      return this.topicMapping['belongs to completed goals'];
+    } else {
+      return this.topicMapping['belongs to neither'];
+    }
+  }
+
+  getStaticImageUrl(imagePath: string): string {
+    return this.urlInterpolationService.getStaticImageUrl(imagePath);
+  }
+
+  topicBelongToLearntToPartiallyLearntTopics(topicName: string): boolean {
+    if (this.learntToPartiallyLearntTopics.includes(topicName)) {
+      return true;
+    }
+    return false;
+  }
+
+  toggleStory(index: string): void {
+    this.currentGoalsStoryIsShown[index] = !(
+      this.currentGoalsStoryIsShown[index]);
+  }
+
+  addToLearnerGoals(
+      topic: LearnerTopicSummary, topicId: string, index: number): void {
+    var activityId = topicId;
+    var activityType = constants.ACTIVITY_TYPE_LEARN_TOPIC;
+    if (!this.topicIdsInCurrentGoals.includes(activityId)) {
+      var isSuccessfullyAdded = (
+        this.learnerDashboardActivityBackendApiService.addToLearnerGoals(
+          activityId, activityType));
```

Comment:
It's because the topic addition depends on 3 cases: Goals limit exceeded, Topic belongs to learnt goals and successfully added. We will perform the thing below only when the result is successfully added.

Reply:
Those checks should be there in the frontend as well. Also, if you are waiting on response from the backend, why isn't it inside a promise?

Reply:
Done. PTAL!